src/utils/utils_log.c

```diff
@@ -0,0 +1,204 @@
+    switch (l) {
+    case LOG_DEBUG:
+        return "DEBUG";
+    case LOG_ERROR:
+        return "ERROR";
+    case LOG_INFO:
+        return "INFO";
+    case LOG_WARNING:
+        return "WARN";
+    default:
+        ASSERT(0);
+    }
+}
+static const char *matchEnvVar(const char *envVar, const char *prefix,
+                               const char **extraArg) {
+    const char *found = strstr(envVar, prefix);
+    // ensure that found string is a first on list or it's separated semicolon
+    if (!found) {
+        return NULL;
+    }
+    // if found string is not at the begining of envVar ensure if it's preceded by ';'
+    if (found != envVar && *(found - 1) != ';') {
+        return NULL;
+    }
+
+    const char *endFound = found + strlen(prefix);
+    if (!extraArg) {
+        // if there is no argument matched string should end with ';' or '\0'
+        if (*endFound != '\0' && *endFound != ';') {
+            return NULL;
+        }
+        return found;
+    }
+
+    // check if matched string ends with '='
+    if (*endFound != '=') {
+        return NULL;
+    }
+
+    *extraArg = endFound + 1;
+
+    return found;
+}
+
+void utils_log(LogLevel level, const char *format, ...) {
+
+    if (!loggerConfig.output) {
+        loggerConfig.output = stderr;
+    }
+    if (level < loggerConfig.level) {
+        return;
+    }
+
+#ifdef _WIN32
+    DWORD pid = GetCurrentProcessId();
+    DWORD tid = GetCurrentThreadId();
+#else
+    pid_t pid = getpid();
+    pid_t tid = gettid();
+#endif
+
+    char buffer[2048];
+    va_list args;
+    va_start(args, format);
+    int ret = vsnprintf(buffer, sizeof(buffer), format, args);
+    const char *overflow = "";
+    if (ret >= (intptr_t)sizeof(buffer)) {
+        //TODO: alloc bigger buffer with base alloc
+        overflow = "[truncated...]";
+    }
+    va_end(args);
+    char header[256];
+    char *h_pos = header;
+    memset(header, 0, sizeof(header));
+    *h_pos++ = '[';
+
+    if (loggerConfig.timestamp) {
+        time_t now = time(NULL);
+        struct tm tm_info;
+        localtime_r(&now, &tm_info);
+
+        ASSERT((intptr_t)sizeof(header) >
+               (h_pos - header)); //should never happen
+        h_pos += strftime(h_pos, sizeof(header) - (h_pos - header),
+                          "%Y-%m-%dT%H:%M:%S ", &tm_info);
+    }
+
+    if (loggerConfig.pid) {
+        ASSERT((intptr_t)sizeof(header) >
+               (h_pos - header)); //should never happen
+        h_pos += snprintf(h_pos, sizeof(header) - (h_pos - header),
+                          "PID:%-6lu TID:%-6lu ", (unsigned long)pid,
+                          (unsigned long)tid);
+    }
+
+    char logLine[2560];
+    snprintf(logLine, sizeof(logLine), "%s%-5s UMF] %s%s\n", header,
+             levelToStr(level), buffer, overflow);
+
+    fputs(logLine, loggerConfig.output);
+
+    if (level >= loggerConfig.flushLevel) {
+        fflush(loggerConfig.output);
+    }
+}
+
+#define MAX_FILE_PATH 256
```

Comment:
are there any os-specific functions to get this?

Reply:
maybe - but not compile time. Also it's more our limitation, so we can handle this easier (and without allocations).

Reply:
Please move all definitions at the top of the file. This is the first place were I look for them. Otherwise it requires ctrl+f to find the values.